### PR TITLE
GEOMESA-1986,GEOMESA-1987,GEOMESA-1993 IDL Safe TubeSelect, Interpolated GapFill, Binning Fix 

### PR DESCRIPTION
**Pull request by jmglenn**
This PR combines changes from the previous PR#1659 as well as the following updates:
GEOMESA-1986 TubeSelect Modified to Properly Handle IDL by segmenting lines crossing the IDL into two lines with proper times that stop at the IDL.
GEOMESA-1987 Fixed binning behavior to be more intuitive. If maxBins is 0 then don't bin the features, if 1 then make one bin, otherwise calculate number of bins based on numFeatures and maxBins
GEOMESA-1993 Added an interpolated gap fill which divides longer gaps into smaller space time interpolated segments based on the passed in buffer size.

### CODE REVIEW

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 1]
```diff
@@ -97,6 +97,55 @@ abstract class TubeBuilder(val tubeFeatures: SimpleFeatureCollection,
...
   }
 
+  /**
+    * This function checks if a segment crosses the IDL.
+    * @param point1 The first point in the segment
+    * @param point2 The second point in the segment
+    * @return boolean true if the segment crosses the IDL, otherwise false
+    */
+  def crossesIDL(point1:Coordinate,point2:Coordinate): Boolean = {
+    ((point1.x > 0 && point2.x < 0) || (point2.x > 0 && point1.x < 0))
```
Comment (elahrvivaz) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: wouldn't this also catch points along the prime meridian? I believe the standard check done in JTS is if the points are more than 180 degrees apart.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 2]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+
+      // result set to tube on
+      val features = fs.getFeatures(CQL.toFilter("type = 'pool' or type = 'pool2'"))
+
+      // get back type b from tube
+      val ts = new TubeSelectProcess()
+      val startTS = System.currentTimeMillis()
+      val results = ts.execute(tubeFeatures, features, null, 1L, 1L, 1000.0, 5, "interpolated")
+      val endTS = System.currentTimeMillis()
+      println("Tube Select Duration: " + String.valueOf((endTS-startTS)/1000))
```
Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: remove println or convert to logger

**Comment by jahhulbert-ccri** [comment]
When we squash let's change the commit msg to say "GEOMESA-1986 GEOMESA-1987 GEOMESA-1993"

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 3]
```diff
@@ -97,6 +97,55 @@ abstract class TubeBuilder(val tubeFeatures: SimpleFeatureCollection,
     }
   }
 
+  /**
+    * This function checks if a segment crosses the IDL.
+    * @param point1 The first point in the segment
+    * @param point2 The second point in the segment
+    * @return boolean true if the segment crosses the IDL, otherwise false
+    */
+  def crossesIDL(point1:Coordinate,point2:Coordinate): Boolean = {
```
Comment (jahhulbert-ccri) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: [picky] space after the comma

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 4]
```diff
@@ -144,6 +196,9 @@ class NoGapFill(tubeFeatures: SimpleFeatureCollection,
   }
 }
 
+
```
Comment (jahhulbert-ccri) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: remove extra newlines

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 5]
```diff
@@ -97,6 +97,55 @@ abstract class TubeBuilder(val tubeFeatures: SimpleFeatureCollection,
...
+      //Find the latitude where the segment intercepts the IDL
+      val latIntercept = calcIDLIntercept(input1,input2)
+      val p1 = new Coordinate(-180,latIntercept)
+      val p2 = new Coordinate(180,latIntercept)
+      //This orders the points so that point1 is always the east-most point
+      val (point1, point2) = if(input1.x > 0) (input1, input2) else (input2, input1)
+      val westLine = new LineString(new CoordinateArraySequence(Array(p1, point2)), geoFac)
+      val eastLine = new LineString(new CoordinateArraySequence(Array(point1, p2)), geoFac)
+      new MultiLineString(Array[LineString](westLine,eastLine), geoFac)
+    }else{
```
Comment (jahhulbert-ccri) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: [style] space after curly ```} else{```

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala (hunk 1 above)
Comment (jahhulbert-ccri) [reply geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: yeah we may also want to make sure the unit test tests this with a line directly on it?

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 6]
```diff
@@ -97,6 +97,55 @@ abstract class TubeBuilder(val tubeFeatures: SimpleFeatureCollection,
...
+
+  /**
+    * Return an Array containing either 1 or 2 LineStrings that straddle but
+    * do not cross the IDL.
+    * @param input1 The first point in the segment
+    * @param input2 The second point in the segment
+    * @return an array of LineString containing either 1 or 2 LineStrings that do not
+    *         span the IDL.
+    */
+  def makeIDLSafeLineString(input1:Coordinate, input2:Coordinate): Geometry = {
```
Comment (jahhulbert-ccri) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: eventually we'll move this into a until class perhaps?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 7]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
 
       results.size mustEqual 20
     }
+
+    "should properly handle geometries crossing the IDL" in {
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: `<curmudgeon_mode>`I know that nobody cares, but this is NOT a fix for the IDL (a complicated line segment); it is a fix for the anti-meridian.`</curmudgeon_mode>`

Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: plus, 'anti-meridian' is a cool word

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 8]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+      // write the feature to the store
+      val res = fs.addFeatures(featureCollection)
+
+      // tube features
+      val tubeFeatures = fs.getFeatures(CQL.toFilter("type = 'c'"))
+
+      // result set to tube on
+      val features = fs.getFeatures(CQL.toFilter("type <> 'c'"))
+
+      // get back type b from tube
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: [picky] Does the comment mean `type b` or `type d`?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 7 above)
Comment (jahhulbert-ccri) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: we need tshirts

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 9]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+
+      ds.createSchema(sft)
+      val fs = ds.getFeatureSource(sftName)
+
+      val featureCollection = new DefaultFeatureCollection(sftName, sft)
+      //TODO: Make data that actually emulated a track, and two alternate tracks, one offset from the other by a small amount of time
+      val calc = new GeodeticCalculator()
+      //Configure track generation for 1 hour at 10.2 meters per second @ 80 degrees starting at -141.0, 35.0
+      var curPoint = WKTUtils.read("POINT(-141.0 35.0)")
+      val timePerPoint = 150000l  //2.5 minutes in milliseconds
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: [style] The `l` might be easier to read as `L` when adjacent to other digits.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 10]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+      val speed =.0102d // meters/millisecond
+      val startTime = new DateTime(f"2011-01-01T00:00:00Z", DateTimeZone.UTC).toInstant.getMillis
+      val endTime = new DateTime(f"2011-01-01T01:00:00Z", DateTimeZone.UTC).toInstant.getMillis
+      val heading = 80.0d
+      //Validation values
+      var i=0                     //Index for providing an FID to each point
+      var trackCount = 0          //Count number of track points
+      var poolCount = 0           //Count number of valid pool points
+      var invalidPoolCount = 0    //Count number of invalid pool points (those back in time)
+
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: [style] Extraneous blank line.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 11]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+      ds.createSchema(sft)
+      val fs = ds.getFeatureSource(sftName)
+
+      val featureCollection = new DefaultFeatureCollection(sftName, sft)
+      //TODO: Make data that actually emulated a track, and two alternate tracks, one offset from the other by a small amount of time
+      val calc = new GeodeticCalculator()
+      //Configure track generation for 1 hour at 10.2 meters per second @ 80 degrees starting at -141.0, 35.0
+      var curPoint = WKTUtils.read("POINT(-141.0 35.0)")
+      val timePerPoint = 150000l  //2.5 minutes in milliseconds
+      val speed =.0102d // meters/millisecond
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: [style] The `d` is extraneous here.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 12]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+        i+=1
+        val sf = AvroSimpleFeatureFactory.buildAvroFeature(sft,List(), i.toString)
+        sf.setDefaultGeometry(curPoint)
+        sf.setAttribute(DefaultDtgField, new DateTime(t, DateTimeZone.UTC).toDate)
+        sf.getUserData()(Hints.USE_PROVIDED_FID) = java.lang.Boolean.TRUE
+        featureCollection.add(sf)
+        if(t % 600000 == 0 ){         //Create a track point every 10 minutes
+          sf.setAttribute("type","track")
+          trackCount+=1
+        } else if (t % 450000 == 0) { //Create invalid pool points every 7 minutes, but put them back in time.
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: [picky] Isn't 450,000 milliseconds actually 7.5 minutes?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala [hunk 13]
```diff
@@ -259,8 +262,141 @@ class TubeSelectProcessTest extends Specification {
...
+
+
+      (startTime to endTime).by(timePerPoint).foreach { t =>
+        i+=1
+        val sf = AvroSimpleFeatureFactory.buildAvroFeature(sft,List(), i.toString)
+        sf.setDefaultGeometry(curPoint)
+        sf.setAttribute(DefaultDtgField, new DateTime(t, DateTimeZone.UTC).toDate)
+        sf.getUserData()(Hints.USE_PROVIDED_FID) = java.lang.Boolean.TRUE
+        featureCollection.add(sf)
+        if(t % 600000 == 0 ){         //Create a track point every 10 minutes
```
Comment (cne1x) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: Did you intend to use `(t - startTime) % 600000`?  Not that it makes any major difference...

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 7 above)
Comment (jmglenn) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: So should I change all instances of IDL to read AntiMeridian or AM for accuracy's sake?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 8 above)
Comment (jmglenn) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: Thank you for catching that. Fixed. It should be d.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 9 above)
Comment (jmglenn) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: I'll remember that. Thank you. Fixed.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 12 above)
Comment (jmglenn) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: Yes. I'll update the comment.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 13 above)
Comment (jmglenn) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: Actually, good catch. When I read your comment I was thinking about it and if startTime was not itself divisible by 600000 then t would not reliably be either at the desired intervals. I will update that. I think I lucked out with my chosen starting time and that problem didn't come up.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala (hunk 11 above)
Comment (jmglenn) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/process/tube/TubeSelectProcessTest.scala]: Fixed

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 14]
```diff
@@ -97,6 +97,55 @@ abstract class TubeBuilder(val tubeFeatures: SimpleFeatureCollection,
...
+
+  /**
+    * Calculate the latitude at which the segment intercepts the IDL.
+    * @param point1 The first point in the segment
+    * @param point2 The second point in the segment
+    * @return a dobule representing the intercept latitude
+    */
+  def calcIDLIntercept(point1: Coordinate, point2: Coordinate): Double = {
+    if(point2.x > point1.x){
+      point2.y - (((point2.y-point1.y)/(point2.x-(-180+point1.x)))*point2.x)
```
Comment (cne1x) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: Okay, I'm probably just doing the algebra wrong, but when I put in the proportions, I don't get this formula.  Instead, I get something like:  `intercept = y2 - (y2 - y1)(x2 - 180)/(x2 - x1)`  Clearly, using 180 as the X-value where the segment between the two points intercepts the anti-meridian.  Please let me know what I've mistaken.  Thanks!

**Comment by elahrvivaz** [comment]
@jmglenn have all comments been addressed? is this ready to go?

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala [hunk 15]
```diff
@@ -97,6 +97,55 @@ abstract class TubeBuilder(val tubeFeatures: SimpleFeatureCollection,
     }
   }
 
+  /**
+    * This function checks if a segment crosses the IDL.
+    * @param point1 The first point in the segment
+    * @param point2 The second point in the segment
+    * @return boolean true if the segment crosses the IDL, otherwise false
+    */
+  def crossesIDL(point1:Coordinate, point2:Coordinate): Boolean = {
```
Comment (anthonyccri) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/tube/TubeBuilder.scala]: These are all generally useful utilities.  Please move them into the `geomesa-util` module and give them some specific tests.  Also, please see my email for a slight change to the computation.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeometryUtils.scala [hunk 16]
```diff
@@ -105,4 +105,41 @@ object GeometryUtils {
...
   private def allRightAngles(p: Polygon): Boolean =
     p.getCoordinates.sliding(2).forall { case Array(left, right) => left.x == right.x || left.y == right.y }
+
+  /**
+    * This function checks if a segment crosses the IDL.
+    * @param point1 The first point in the segment
+    * @param point2 The second point in the segment
+    * @return boolean true if the segment crosses the IDL, otherwise false
+    */
+  def crossesIDL(point1:Coordinate, point2:Coordinate): Boolean = {
```
Comment (anthonyccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeometryUtils.scala]: Ok, last comment.  We need specific tests for these methods in `GeometryUtilsTest`.

**Comment by elahrvivaz** [comment]
@jmglenn ready for merge?

**Pull request merged**